File: Focus/Focus/Domain/NodeMetadata.cs
#nullable enable

using System;
using System.Collections.Generic;

namespace Systems.Sanity.Focus.Domain;

public sealed class NodeMetadata
{
    public DateTimeOffset CreatedAtUtc { get; set; }

    public DateTimeOffset UpdatedAtUtc { get; set; }

    public string? Source { get; set; }

    public string? Device { get; set; }

    public List<NodeAttachment> Attachments { get; set; } = new();

    public static NodeMetadata Create(
        string? source,
        string? device,
        DateTimeOffset? timestampUtc = null)
    {
        var normalizedTimestamp = NormalizeTimestamp(timestampUtc);
        return new NodeMetadata
        {
            CreatedAtUtc = normalizedTimestamp,
            UpdatedAtUtc = normalizedTimestamp,
            Source = source,
            Device = device,
            Attachments = new()
        };
    }

    public void Touch(DateTimeOffset? timestampUtc = null)
    {
        UpdatedAtUtc = NormalizeTimestamp(timestampUtc);
    }

    private static DateTimeOffset NormalizeTimestamp(DateTimeOffset? timestampUtc) =>
        (timestampUtc ?? DateTimeOffset.UtcNow).ToUniversalTime();
}

public sealed class NodeAttachment
{
    public Guid Id { get; set; }

    public string RelativePath { get; set; } = string.Empty;

    public string MediaType { get; set; } = string.Empty;

    public string DisplayName { get; set; } = string.Empty;

    public DateTimeOffset CreatedAtUtc { get; set; }
}

File: android/app/src/main/java/com/systemssanity/focus/domain/model/NodeMetadata.kt
package com.systemssanity.focus.domain.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

object NodeMetadataSources {
    const val Manual = "manual"
    const val ClipboardText = "clipboard-text"
    const val ClipboardImage = "clipboard-image"
    const val LegacyImport = "legacy-import"
}

@Serializable
data class NodeAttachment(
    @SerialName("id") val id: String = "",
    @SerialName("relativePath") val relativePath: String = "",
    @SerialName("mediaType") val mediaType: String = "",
    @SerialName("displayName") val displayName: String = "",
    @SerialName("createdAtUtc") val createdAtUtc: String = ClockProvider.nowIsoSeconds(),
)

@Serializable
data class NodeMetadata(
    @SerialName("createdAtUtc") val createdAtUtc: String = ClockProvider.nowIsoSeconds(),
    @SerialName("updatedAtUtc") val updatedAtUtc: String = createdAtUtc,
    @SerialName("source") val source: String? = NodeMetadataSources.Manual,
    @SerialName("device") val device: String? = DeviceIdentity.defaultDeviceName,
    @SerialName("attachments") val attachments: List<NodeAttachment> = emptyList(),
) {
    fun touched(timestamp: String = ClockProvider.nowIsoSeconds()): NodeMetadata =
        copy(updatedAtUtc = timestamp)
}
